### ISSUE TITLE
Google Drive: No way to delete orphan files

### ISSUE THREAD

**Issue by Zeioth** [rclone/rclone#4166]
On Google drive, I deleted a directory that was created by rclone, and now all the 40.000 files inside of it are orphan. There's some way to use RClone to delete all orphan files on my drive? Thank you.

### What is your rclone version (output from rclone version)
Latest (AUR) -> rclone v1.51.0

### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Manjaro

### Which cloud storage system are you using? (eg Google Drive)
Google drive

**Comment by Zeioth** [comment]
Possible solution:
https://medium.com/@staticmukesh/how-i-managed-to-delete-1-billions-files-from-my-google-drive-account-9fdc67e6aaca

**Comment by animosity22** [comment]
@ncw - I know we've looked before and I still do not see any API way to do this rather than that longer complex way listed a few posts online like the one above. I can't see this coming into rclone unless they expose the API so I can mark an as an enhancement for Drive and let me know what you want to do with it.

**Comment by Zeioth** [comment]
I think now it's possible. Look, try [this query](https://developers.google.com/drive/api/v3/reference/files/list?apix_params=%7B%22pageSize%22%3A100%2C%22q%22%3A%22%27root%27%20in%20parents%20%22%7D) with 

`q='root' in parents`

**Comment by Zeioth** [comment]
For some stupid reason, I need to validate my OAuth  credentials in order to delete files from Google Drive, which apparently takes 4-6 weeks, so I can't confirm that everything works yet, but this should delete only orphan files (not sure about directories):

https://pastebin.com/a4ySAJ69

**The premise is:**

* List all files.
* If a file has no 'parents' field, it means it's an orphan file.
* So, the script deletes them.

Feedback is welcome.

**Comment by ncw** [comment]
This appears to be quite tricky...

To make it work in rclone we'd need to find [a query](https://developers.google.com/drive/api/v3/search-files) which meant this file has no parents - I'm not sure there is one though...

Assuming not, then you'd have to do the initial query with no `... in parents` filter which would mean that you'd get **all** items returned. You'd then manually discard any item with parents.

That would work, but it would be very slow for drives with lots of files.

If anyone wants to have a go at a `--drive-orphans` flag, here is the code you need to be looking at (showing the trashed only function)

https://github.com/rclone/rclone/blob/8e91f83174cc4011158c596d0c8569cd226fc36d/backend/drive/drive.go#L679-L681

**Comment by ncw** [comment]
> On Google drive, I deleted a directory that was created by rclone, and now all the 40.000 files inside of it are orphan. There's some way to use RClone to delete all orphan files on my drive? 

How did you delete that directory? Was it through rclone? If so we need to fix that!

**Comment by Zeioth** [comment]
No it wasn't through RClone: I deleted the directory created by RClone, manually on Google Drive website. Apparently when you do that, it generates orphan files. It's on Google Drive documentation.

It's their fault, not RClone.

**Comment by ncw** [comment]
> No it wasn't through RClone: I deleted the directory created by RClone, manually on Google Drive website. Apparently when you do that, it generates orphan files. It's on Google Drive documentation.

Really! That is weird!

**Comment by ncw** [comment]
I've added the framework for backend commands now, so if someone wanted to make a "rescue-orphans" command for the drive backend that would now be possible.

**Comment by Ret2lib** [comment]
Hi,

I'm willing to help contribute to this but not sure what is meant by "framework for backend commands". Is there somewhere I can read about on this? What part of the code do I need to add this to (I just found out about rclone today)

**Comment by ncw** [comment]
@Ret2lib sorry for the delay responding.

The backend commands are defined here

https://github.com/rclone/rclone/blob/654f5309b041cbb65bba3b86423fcb75c0d84dab/backend/drive/drive.go#L3077

And implemented here

https://github.com/rclone/rclone/blob/654f5309b041cbb65bba3b86423fcb75c0d84dab/backend/drive/drive.go#L3211

So you'd make a function which did the work and parse the arguments in the `Command` function, then call your function.

You'd use it something like this

    rclone backend rescueorphans drive:dir-to-rescue-files

Let me know if you'd like more help and I'll provide it :-)

**Comment by SaladinoGW** [comment]
> @Ret2lib sorry for the delay responding.
> 
> The backend commands are defined here
> 
> https://github.com/rclone/rclone/blob/654f5309b041cbb65bba3b86423fcb75c0d84dab/backend/drive/drive.go#L3077
> 
> And implemented here
> 
> https://github.com/rclone/rclone/blob/654f5309b041cbb65bba3b86423fcb75c0d84dab/backend/drive/drive.go#L3211
> 
> So you'd make a function which did the work and parse the arguments in the `Command` function, then call your function.
> 
> You'd use it something like this
> 
> ```
> rclone backend rescueorphans drive:dir-to-rescue-files
> ```
> 
> Let me know if you'd like more help and I'll provide it :-)

I just lost 700GB of space at this point because the files became orphaned when I deleted them via the web, I had sent them via rclone copy to a shared directory. is:unorganized owner:me via the web can't find it because there are too many files unfortunately. Is there any way you can implement it please?

**Comment by dzg** [comment]
Did anything ever get solved here? I have millions of orphaned files.

**Comment by ncw** [comment]
@dzg I had a go at implementing this

[v1.68.0-beta.8105.96f964bbb.fix-4166-drive-rescue-orphans](https://beta.rclone.org/branch/fix-4166-drive-rescue-orphans/v1.68.0-beta.8105.96f964bbb.fix-4166-drive-rescue-orphans/) on branch [fix-4166-drive-rescue-orphans](https://github.com/rclone/rclone/tree/fix-4166-drive-rescue-orphans) (uploaded in 15-30 mins)

Use like this

### rescue

Rescue or delete any orphaned files

    rclone backend rescue remote: [options] [<arguments>+]

This command rescues or deletes any orphaned files or directories.

Sometimes files can get orphaned in Google Drive. This means that they
are no longer in any folder in Google Drive.

This command finds those files and either rescues them to a directory
you specify or deletes them.

Usage:

This can be used in 3 ways.

First, list all orphaned files

    rclone backend rescue drive:directory

Second rescue all orphaned files to the directory indicated

    rclone backend rescue drive:directory "relative/path/to/rescue/directory"

Third delete all orphaned files to the trash

    rclone backend rescue drive:directory -o delete


**HOWEVER** this does not work for me. It finds the rescued files just fine, but trying to delete them gives 403 permission denied, and trying to rescue them into a directory (add a parent) gives "Increasing the number of parents is not allowed, cannotAddParent". I tried lots of variations of adding/removing parents but I couldn't get anything to work.

So I think that these files really are orphaned. The API no longer lets us add a parent to rescue them, and the API doesn't seem to let us delete them either.

It would be worth having a try with it though to see if you have the same experience.

**Comment by timwsuqld** [comment]
@ncw I had a quick play with your beta version, because I'm tackling this issue in my own tooling at the moment. Unfortunately, rclone with your mods is detecting more orphaned files/directories than my system does. Investigating a bit more, it appears to be detecting files/directories owned by other users (but that the current user has access to), as being orphaned as well.

The way I had implemented it in my tools was I had to query all files with the query `'me' in owners` and then check if `parents` was defined.

So because rclone is detecting more, I can't test the "rescue" part of this.

But I think we are in the same boat, what used to work (adding a new parent) no longer works. I know it used to work, because I used this code 6-10 months ago and it was working then. Now I get an error no matter what combination I try. Specifying an empty `removeParents`, not specifying `removeParents`, specifying 'root' for `removeParents`, trying the MyDrive ID in `removeParents`. The only request that doesn't error, is if I try the destination folder in `removeParents` and `addParents`, but then it gives me an OK without actually moving it.

```
googleapiclient.errors.HttpError: <HttpError 500 when requesting https://www.googleapis.com/drive/v3/files/<ORHPANED FILE>?addParents=<DESTINATION FOLDER>&removeParents=&alt=json returned "Internal Error". Details: "[{'message': 'Internal Error', 'domain': 'global', 'reason': 'internalError'}]">

googleapiclient.errors.HttpError: <HttpError 500 when requesting https://www.googleapis.com/drive/v3/files/<ORHPANED FILE>?addParents=<DESTINATION FOLDER>&alt=json returned "Internal Error". Details: "[{'message': 'Internal Error', 'domain': 'global', 'reason': 'internalError'}]">

etc
```

So I'm pretty sure something has changed in the API, and we're just stuffed now. I'll update you if I find any API call that works for rescuing the orphaned files/fodlers

**Comment by timwsuqld** [comment]
@ncw I think I've worked it out. I had 2 users with issues, and both folders had the same name "USB and External Devices". No API calls worked to modify those folders, although I could use rclone to move everything from inside the folder and that worked. It turns out that is a folder from the Google Drive Backup client (when it's backing up your external disks). I'm guessing there are other folders that appear as orphans, but are part of the computer backup part of Google Drive.

I had another user with plenty of orphaned files and folders, and the normal `files.update` call worked (the params I am using are:
```python
update(
     fileId=file["id"],
     addParents=destination_folder_o["id"],
     supportsAllDrives=True,
     supportsTeamDrives=True,
     fields="parents",
```

To be clear, I check in code before executing this that they really do have no parents, if they do have parents (but with another users), then this would fail without the relevant removeParents as well.

In theory, this should be less and less of an issue, you can't orphan files anymore, if you attempt to orphan them, it'll put them into your My Drive.

I will see if I can find any more orphaned files and I'll test your beta version on those to see if it works. (I'm progressively working through 600 users, so I'll find more orphaned files)

**Comment by timwsuqld** [comment]
@ncw I have another user with orphaned files, and I can see the issue you are having. rclone just finds files without parents, but that doesn't mean the file is orphaned. You can have a file shared with you, so it ends up with no parents, but it really does have a parent. Can you modify the code so it only considers a file you own and without parents, as being orphaned. Then I think I can test it.

Your 403 is because you are trying to add a parent to a file owned by someone else. Google no longer allows files to have multiple parents, so even though you can't see the parent (as you don't have permission to the parent), you can't add another parent, and need to instead create a shortcut.

I did try with `--drive-auth-owner-only` but that doesn't seem to make a difference to the backend command.

**Comment by ncw** [comment]
Thanks for testing @timwsuqld 

It looks like you are right this needs to check if files are owned by the user. I did this by adding `'me' in owners` to the query which seems to work.

I used this to rescue some orphans from my own drive which were files I created in a shared drive that have been deleted (I think) from the shared drive.

Please give this a go

[v1.69.0-beta.8368.01bc25464.fix-4166-drive-rescue-orphans](https://beta.rclone.org/branch/fix-4166-drive-rescue-orphans/v1.69.0-beta.8368.01bc25464.fix-4166-drive-rescue-orphans/) on branch [fix-4166-drive-rescue-orphans](https://github.com/rclone/rclone/tree/fix-4166-drive-rescue-orphans) (uploaded in 15-30 mins)

**Comment by ncw** [comment]
Excellent news :-)

> I did trip on the syntax. I did `rclone backend rescue drive: "./directory"` which of course made a folder named `.` in the root. 

Ha! Yes I did something very similar when testing and made a directory called `drive:Orphans`!

> It is a bit of a strange syntax, especially because you stated `rclone backend rescue drive:directory`, but I'm pretty sure the directory means nothing, because an orphaned file by it's nature has no directory as a parent, so I think that's what tripped me. I just probably need to look at the example better.

Hmm. Yes you are right the `directory` in `drive:directory` is not used and is confusing in the help.

I tweaked the instructions below - any more improvements?

Do you think this feature is ready to merge now?

Thanks

Nick

Version with tweaked help but otherwise identical

[v1.69.0-beta.8368.97fcf89c1.fix-4166-drive-rescue-orphans](https://beta.rclone.org/branch/fix-4166-drive-rescue-orphans/v1.69.0-beta.8368.97fcf89c1.fix-4166-drive-rescue-orphans/) on branch [fix-4166-drive-rescue-orphans](https://github.com/rclone/rclone/tree/fix-4166-drive-rescue-orphans) (uploaded in 15-30 mins)

### rescue

Rescue or delete any orphaned files

    rclone backend rescue remote: [options] [<arguments>+]

This command rescues or deletes any orphaned files or directories.

Sometimes files can get orphaned in Google Drive. This means that they
are no longer in any folder in Google Drive.

This command finds those files and either rescues them to a directory
you specify or deletes them.

Usage:

This can be used in 3 ways.

First, list all orphaned files

    rclone backend rescue drive:

Second rescue all orphaned files to the directory indicated

    rclone backend rescue drive: "relative/path/to/rescue/directory"

e.g. To rescue all orphans to a directory called "Orphans" in the top level

    rclone backend rescue drive: Orphans

Third delete all orphaned files to the trash

    rclone backend rescue drive: -o delete

**Comment by timwsuqld** [comment]
I think those updated instructions should hopefully be good enough. I think this can be merged now. Thanks heaps!

**Comment by ncw** [comment]
Thank you for testing and your very helpful comments in getting this working properly @timwsuqld 

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.69
